pocs/camera/zwo.py

```diff
@@ -0,0 +1,446 @@
+        readout_thread.start()
+
+    def _readout(self, filename, width, height, header, exposure_event):
+        timer = CountdownTimer(duration=self._timeout)
+        try:
+            exposure_status = Camera._ASIDriver.get_exposure_status(self._camera_ID)
+            while exposure_status == 'WORKING':
+                if timer.expired():
+                    msg = "Timeout waiting for exposure on {} to complete".format(self)
+                    raise error.Timeout(msg)
+                time.sleep(0.01)
+                exposure_status = Camera._ASIDriver.get_exposure_status(self._camera_ID)
+        except RuntimeError as err:
+            # Error returned by driver at some point while polling
+            self.logger.error('Error while waiting for exposure on {}: {}'.format(self, err))
+            raise err
+        else:
+            if exposure_status == 'SUCCESS':
+                try:
+                    image_data = Camera._ASIDriver.get_exposure_data(self._camera_ID,
+                                                                     width,
+                                                                     height,
+                                                                     self.image_type)
+                except RuntimeError as err:
+                    raise error.PanError('Error getting image data from {}: {}'.format(self, err))
+                else:
+                    fits_utils.write_fits(image_data, header, filename, self.logger, exposure_event)
+            elif exposure_status == 'FAILED':
+                raise error.PanError("Exposure failed on {}".format(self))
+            elif exposure_status == 'IDLE':
+                raise error.PanError("Exposure missing on {}".format(self))
+            else:
+                raise error.PanError("Unepected exposure status on {}: '{}'".format(
+                    self, exposure_status))
+        finally:
+            exposure_event.set()  # write_fits will have already set this, *if* it got called.
+            self._exposure_lock.release()
+
+    def _fits_header(self, seconds, dark):
+        header = super()._fits_header(seconds, dark)
+        header.set('CAM-GAIN', self.gain, 'Internal units')
+        header.set('CAM-BITS', int(get_quantity_value(self._info['bit_depth'], u.bit)),
+                   'ADC bit depth')
+        header.set('XPIXSZ', get_quantity_value(self._info['pixel_size'], u.um), 'Microns')
+        header.set('YPIXSZ', get_quantity_value(self._info['pixel_size'], u.um), 'Microns')
+        header.set('EGAIN', get_quantity_value(self.egain, u.electron / u.adu), 'Electrons/ADU')
+        return header
+
+    def _refresh_info(self):
+        self._info = Camera._ASIDriver.get_camera_property(self._camera_index)
+
+    def _control_getter(self, control_type):
+        if control_type in self._control_info.keys():
+            return Camera._ASIDriver.get_control_value(self._camera_ID, control_type)
```

Comment:
Everywhere you use this above you then return the first element of a list. Does it make sense to just do that here?

Reply:
Maybe... `ASIDriver.get_control_value()` returns both the current value and a boolean indicating whether the value is currently being automatically adjusted by the camera.  For most of the camera parameters auto is not an option so this is usually `False`, and even when auto is on you may not care.  But there probably are some situations where you would want to know so I've left the second value in this private method in case I need it at some point.